docs/src/token.md

```diff
@@ -449,3 +450,134 @@ have a balance of zero to be closed.
 
 ### Non-Fungible tokens
 An NTF is simply a token type where only a single token has been minted.
+
+### Program updates
+
+As the SPL Token program evolves the accounts owned by older version may need to
+be migrated to newer versions.  This can be done by proxy via a migration
+program.  For more information on program updates in general, refer to
+[program updates](program-updates.md).
+
+The following is an example of how a wallet may migrate accounts owned by SPL
+Token to accounts owned by SPL Token v3.
+
+Updating the accounts will require a migration program that serves as an
```

Comment:
Here's the instructions that I'm thinking of for the v2 -> v3 token upgrade program.  The minter of the v3 token assigns a batch of v3 tokens to the program with `OpenReserveAccount`.  It may need to add more v3 tokens as the users upgrade.

Each user, or their wallet, uses `Upgrade`.

The minter can reclaim any remaining v3 tokens from the program with `CloseReserveAccount`.

```rust
    /// Create a v3 token account to hold funds used when upgrading v2 token accounts (if
    /// necessary), and transfer `amount` tokens into it.
    ///
    ///   0. `[writeable]` v3 token account program address
    ///   1. `[]` The v3 mint this account will be associated with.
    ///   2. `[writeable,signer]` The system account that will fund the v3 token account
    ///   3. `[]` Rent sysvar
    ///   4. `[writeable]` funding v3 token account program address.  `amount` tokens will be
    ///                    transferred from this account
    ///   5. `[signer]` owner of the v3 token account (TODO: support multisig owners too...)
    ///
    OpenReserveAccount { amount: u64 },

    /// Close the v3 token account held by the program to upgrade v2 token accounts
    ///
    ///   0. `[writeable]` v3 token account program address
    ///   1. `[]` The v3 mint this account is associated with.
    ///   2. `[writeable]` v3 token account address that will receive any remaining tokens in the
    ///                    program's token account
    ///   3. `[writeable]` The system account that will receive the rent lamports after
    ///                    closing the account
    ///
    CloseReserveAccount,

    /// Update the source v2 token account into the destination v3 token account.  All tokens in
    /// the v2 account will be burned, and equivalent tokens will be credited to the v3 account
    /// from the program's v3 token reserve.
    ///
    /// It's the caller's responsibility to close the v2 account after this instruction completes
    /// if they wish.
    ///
    ///   0. `[writeable]` source v2 token account
    ///   1. `[writeable]` destination v3 token account
    ///   2. `[writeable]` v3 token account program address
    ///   3. `[]` The v3 mint this account will be associated with.
    ///   4. `[signer]` owner of v2 token account (TODO: support multisig owners...)
    ///
    Upgrade,
```

Reply:
Splitting the signing authorities (wallet, dex) apart into different instructions simplifies things.  It also complicates some things.

Some questions/comments
- Who creates the new v3 account, if done by `Upgrade` then the lamport funding account must also be a signer.
- If account #2 in `OpenReserveAccount` is used to fund the new v3 token account then that's potentially a lot of lamports locked up.  Transfering the lamports from the v2 to v3 account directly in the `upgrade` instruction eliminates that.
- The `upgrade` instruction would be called by the wallet, they may or may not be able to call `CloseAccount` which means the dex would have to monitor what accounts have been upgraded and perform and clean-up.  The dex would also have to actively monitor the reserve account if it did not fully fund it to account for all upgrades.
- Why Account #3 in `Upgrade`?  Wouldn't this be already be known by looking at the mint associated with the funding reserve account?
- `Upgrade` would need the v2 mint account to `Burn`
- If `Upgrade` creates and initializes the v3 token account then account #1 will have to be a signer and `Upgrade` will also need the v3 token account authority account.
- The wallet would be paying the tx fee for the upgrade

I was thinking that the wallet would be given a template upgrade tx that it would fill in, sign, and send to the dex, the dex would then finishing signing and submit to the chain.  The dex could respond to these upgrade requests directly to refund the funding account, payer, etc...  Upgrade would look like this:

```rust
    ///   0. `[signer]` Payer for any additional account space.  The v2
    ///      account's lamports are transferred to the new v3 account but if the
    ///      v3 account is larger then additional lamports will be required.
    ///      This payer provides those additional lamports.
    ///   1. `[writable]` The v2 account to migrate.
    ///   2. `[writable]` The v2 token mint.
    ///   3. '[writable, signer]' The new v3 account's pubkey.
    ///   4. `[writable]` The v3 token mint.
    ///   5. `[writable]` The v3 funding account
    ///   6. `[signer]` The v3 funding account authority
    ///   7. `[]` Rent sysvar
    ///   8. `[signer]` the V2 token signer (or multiple for multisignature)
```

Reply:
> * Who creates the new v3 account, if done by `Upgrade` then the lamport funding account must also be a signer.

A prior instruction does it, to avoid baking in assumptions about how it's created.  This 

> * If account #2 in `OpenReserveAccount` is used to fund the new v3 token account then that's potentially a lot of lamports locked up.  Transfering the lamports from the v2 to v3 account directly in the `upgrade` instruction eliminates that.

s/lamports/tokens/ but yes.  But this is a better trade-off than requiring the v3 mint authority sign each `Upgrade`, or requiring the v3 mint authority be assigned to the upgrade program itself.

> * The `upgrade` instruction would be called by the wallet, they may or may not be able to call `CloseAccount` which means the dex would have to monitor what accounts have been upgraded and perform and clean-up.  The dex would also have to actively monitor the reserve account if it did not fully fund it to account for all upgrades.

Yep.   

> * Why Account #3 in `Upgrade`?  Wouldn't this be already be known by looking at the mint associated with the funding reserve account?

The funding reserve account is a program account that's generated from provided v3 mint address.  I was thinking this program could support upgrading accounts from any v2 mint to v3, not just one particular mint.

> * `Upgrade` would need the v2 mint account to `Burn`

Ah, ok.   

> * If `Upgrade` creates and initializes the v3 token account then account #1 will have to be a signer and `Upgrade` will also need the v3 token account authority account.

Na, the v3 token account should already exist before invoking `Update`

> * The wallet would be paying the tx fee for the upgrade

It could.  If there's somebody who's motivated to cause an upgrade, they can pay the fee on behalf of the wallet though (with some coordination):
* 3rd party could create the v3 token account for the user
* 3rd party could fund the transaction with the `Upgrade` instruction in it, once the wallet provides the v2 token owner signature


> ... Upgrade would look like this:

This signature is problematic: `3. '[writable, signer]' The new v3 account's pubkey.`.  Can't assume the v3 account will be on the curve, it might be (and probably will be once #626 finishes) a program address

Reply:
> > * Who creates the new v3 account, if done by `Upgrade` then the lamport funding account must also be a signer.
> 
> A prior instruction does it, to avoid baking in assumptions about how it's created.

Yeah, I guess it doesn't have to be a brand new v3 account :-)
> 
> > * If account #2 in `OpenReserveAccount` is used to fund the new v3 token account then that's potentially a lot of lamports locked up.  Transfering the lamports from the v2 to v3 account directly in the `upgrade` instruction eliminates that.
> 
> s/lamports/tokens/ but yes. But this is a better trade-off than requiring the v3 mint authority sign each `Upgrade`, or requiring the v3 mint authority be assigned to the upgrade program itself.

I meant lamports working with the idea that `Upgrade` would create the new v3 account, it would also be funded by the reserver account both for tokens and lamports.
> 
> > * The `upgrade` instruction would be called by the wallet, they may or may not be able to call `CloseAccount` which means the dex would have to monitor what accounts have been upgraded and perform and clean-up.  The dex would also have to actively monitor the reserve account if it did not fully fund it to account for all upgrades.
> 
> Yep.
> 
> > * Why Account #3 in `Upgrade`?  Wouldn't this be already be known by looking at the mint associated with the funding reserve account?
> 
> The funding reserve account is a program account that's generated from provided v3 mint address. I was thinking this program could support upgrading accounts from any v2 mint to v3, not just one particular mint.

There would be a 3rd mint besides the one's reference in the src or dst accounts?

> 
> > * `Upgrade` would need the v2 mint account to `Burn`
> 
> Ah, ok.
> 
> > * If `Upgrade` creates and initializes the v3 token account then account #1 will have to be a signer and `Upgrade` will also need the v3 token account authority account.
> 
> Na, the v3 token account should already exist before invoking `Update`
> 
> > * The wallet would be paying the tx fee for the upgrade
> 
> It could. If there's somebody who's motivated to cause an upgrade, they can pay the fee on behalf of the wallet though (with some coordination):
> 
>     * 3rd party could create the v3 token account for the user
> 
>     * 3rd party could fund the transaction with the `Upgrade` instruction in it, once the wallet provides the v2 token owner signature
> 
> 
> > ... Upgrade would look like this:
> 
> This signature is problematic: `3. '[writable, signer]' The new v3 account's pubkey.`. Can't assume the v3 account will be on the curve, it might be (and probably will be once #626 finishes) a program address

If the v3 account is created before the `Upgrade` instruction then this signer is not required.  

I like the idea of a single instruction upgrade, the lamports flowing from the v2 to v3 account, and then the result of the `Upgrade` being a closed v2 account.  But I guess I was trying to fit too much into one instruction 😉